### PR TITLE
Early stopping on multiple validation metrics

### PR DESCRIPTION
**Pull request by Proyag**
Following a discussion with @snukky , this should let us early stop on multiple validation metrics.
* For example `--validation-metrics cross-entropy bleu --early-stopping 10 20`
* The previous behaviour is still the default, and specifying fewer arguments for `--early-stopping` will only apply it to the corresponding metrics in order: `--validation-metrics cross-entropy bleu --early-stopping 10` is the same as before.
* We can use `0` to not early stop on one of the metrics: `--validation-metrics cross-entropy bleu translation --early-stopping 10 0 20`

### CODE REVIEW

**Comment by emjotde** [comment]
Hi, Proyag,
Does this break backwards-compatibility of config files?

**Comment by snukky** [comment]
@Proyag The backward-compatibility can be checked with our regression tests. 

I think this feature may also require some modifications in loading/saving of [training state](https://github.com/marian-nmt/marian-dev/blob/master/src/training/training_state.h).

**Comment by frankseide** [comment]
"The backward-compatibility can be checked with our regression tests."

No, this is not good enough. It is unrealistic to expect regression tests to completely cover all use cases. If they fail, then obviously something is wrong, but if they pass, we cannot assume that the other use cases were not affected.

Please analyze the code in how far it affects back compat.

**Comment by emjotde** [comment]
Basically test if a config containing
```
early-stopping: 10
```
works. 

I think I have code in there that upcasts single values to lists if lists are expected. We did that for backwards-compatibility of `--lr-warmup M N`.

**Comment by Proyag** [comment]
I already checked that all regression tests passed.
I also took a random model.npz.yml from the passed tests, changed
```
early-stopping:
    - 10
```
to
```
early-stopping: 10
```
and `marian -c model.npz.yml` still works, and in fact, it converts and displays it like the first case. Looks like you already solved this problem :)

**Comment by emjotde** [comment]
@snukky let's review this.

**Comment by emjotde** [comment]
@snukky ping?

**Comment by snukky** [comment]
We were thinking that this could shorten trainings for which the validation cross-entropy continues improving for a very long time, while BLEU stalles for dozens of epochs. The correct solution was better cleaning/filtering of development and training data, which resulted in better BLEU and shorter training.

I am closing the PR. @Proyag Please feel free to complain and re-open if you think that this is still useful.
